WalletWasabi.Tests/UnitTests/Blockchain/TransactionBuilding/BranchAndBoundTests.cs

```diff
@@ -15,106 +16,176 @@ public class BranchAndBoundTests
 		Assert.Equal(new long[] { 17, 10 }, actualSelection);
 	}
 
+	/// <summary>
+	/// Effective bitcoin value received by the payee is more important than
+	/// payer's total costs.
+	/// </summary>
 	[Fact]
-	public void MoreSelection_ExactMatchIsExpensive()
+	public void MoreSelection_ExactMatchIsMoreExpensiveButStillSelected()
 	{
+		using CancellationTokenSource cts = new(TimeSpan.FromSeconds(30));
+
 		long[] inputValues = new long[] { 35, 17, 10, 5, 3, 2 };
 
 		// Make the second input very expensive to spend so that it is not selected (not likely in reality).
-		long[] inputCosts = new long[] { 1, 1000, 1, 1, 1, 1 };
+		long[] inputCosts = new long[] { 1, 10, 1, 1, 1, 1 };
 
 		long target = 27; // Target that we cannot get as a sum of input values.
 
 		BranchAndBound algorithm = new();
 		MoreSelectionStrategy strategy = new(target, inputValues, inputCosts);
-		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
+		_ = algorithm.TryGetMatch(strategy, out _, cts.Token);
+
+		// Selection (35) costs us 35 + 1 = 36.
+		// Selection (17, 10) is actually more expensive: (17 + 10) + (10 + 1) = 38, but
+		// we use that selection sa 27 is exactly the amount a payee expects.
+		long[] actualSelection = strategy.GetBestSelectionFound()!;
+		Assert.NotNull(actualSelection);
+
+		Assert.Equal(new long[] { 17, 10 }, actualSelection);
+	}
+
+	/// <summary>
+	/// The goal of the test is to verify that we prune branches at points when it is clear that
+	/// the sum of remaining coins is less than what we need to hit our target.
+	/// </summary>
+	/// <remarks>
+	/// This is especially important for cases where user has many coins.
+	/// <para>If the optimization is not in place, you should observe that this test does not really finish fast.</para>
+	/// </remarks>
+	[Fact]
+	public void MoreSelection_RemainingAmountOptimization()
+	{
+		using CancellationTokenSource cts = new(TimeSpan.FromSeconds(30));
+
+		List<long> inputValues = new();
+		inputValues.Add(1_000_000);
 
-		Assert.False(wasSuccessful);
-		Assert.Null(selectedCoins);
+		for (int i = 0; i < 1000; i++)
+		{
+			inputValues.Add(1);
+		}
+
+		// All inputs cost the same.
+		long[] inputCosts = inputValues.Select(x => 1L).ToArray();
+
+		long target = 999_999; // Target that we cannot get as a sum of input values.
+
+		BranchAndBound algorithm = new();
+		MoreSelectionStrategy strategy = new(target, inputValues.ToArray(), inputCosts);
+		_ = algorithm.TryGetMatch(strategy, out _, cts.Token);
 
-		// Selection (17, 10) is actually more expensive: 17 + 10 + 1000 + 1 = 1018.
-		// Whereas (35) costs us 35 + 1 = 36.
+		// Assert that we get expected best solution.
 		long[] actualSelection = strategy.GetBestSelectionFound()!;
 		Assert.NotNull(actualSelection);
 
-		Assert.Equal(new long[] { 35 }, actualSelection);
+		Assert.Equal(new long[] { 1_000_000 }, actualSelection);
+	}
+
+	[Fact(Skip = "If the first coin is not selected the test fails to finish. TODO.")]
```

Comment:
Is it because when the algo randomly doesn't select the first coin, the reamining opt kicks in and the algo doesn't try to include that coin? Do I get it right?

Reply:
Yeah, it's because the first coin can either be selected or not. If it is selected, then it finishes fast. 

The issue is that if the first coin is NOT selected, then the algorithm searches the remaining search space (i.e. 1000 coins each of value 1 and it tries to include/exclude those coins which cannot be done in reasonable time). The thing is that when searching those 1000 coins, we incrementally improve best solution but the algorithm does not do the logical inferring - well, I can have at most sum of 1000 and that's not even close to the target. Essentially, we would need another optimization for this -- something like "before starting LesserStrategy, try to take all coins if it is lower <= target".